lib/rules/object-shorthand.js

```diff
@@ -109,7 +109,7 @@ module.exports = {
                 }
 
                 // only computed methods can fail the following checks
-                if (!APPLY_TO_METHODS && node.computed) {
+                if ((!APPLY_TO_METHODS || node.value.type !== "FunctionExpression") && node.computed) {
```

Comment:
I think you can drop the ``APPLY_TO_METHODS` check, the only computed case we care about is `node.value.type === "FunctionExpression" && node.computed`.